src/borg/repository.py

```diff
@@ -705,6 +736,8 @@ def delete(self, id, wait=True):
             segment, offset = self.index.pop(id)
         except KeyError:
             raise self.ObjectNotFound(id, self.path) from None
+        self.shadow_index.setdefault(id, [])
+        self.shadow_index[id].append(segment)
```

Comment:
setdefault also works like a get, so you can do the .append(segment) in the first line.